### ISSUE TITLE
Failure to satisfy dependency range requirement on yarn >=0.27.1

### ISSUE THREAD

**Issue by torgeir** [yarnpkg/yarn#4237]
**Do you want to request a *feature* or report a *bug*?**
Possible regression between 0.27.0 and 0.27.1 (also seen in 0.27.5).

**What is the current behavior?**
Yarn 0.27.1 fails to dedupe package to satisfy the dependencies' ranges (on jquery ^2.1.3 and
pickadate ^3.5.3 (jquery >= 1.7)) and installs two versions of the package:

- jquery 3.2.1
- jquery 2.2.4

**If the current behavior is a bug, please provide the steps to reproduce.**
https://github.com/torgeir/yarn-dedupe-regression-on-0.27.1/blob/master/readme.md

**What is the expected behavior?**
A single version of the package is installed, that satisfies the dependency ranges of both the [project package.json (^2.1.3)](https://github.com/torgeir/yarn-dedupe-regression-on-0.27.1/blob/master/package.json#L10) and the dependency [pickadate's package.json (>=1.7)](https://github.com/amsul/pickadate.js/blob/master/package.json#L39):

- jquery: 2.2.4

**Please mention your node.js, yarn and operating system version.**
node: v8.2.1
yarn: v0.27.1
os: macos sierra 0.12.6 (16G29)

**Comment by BYK** [comment]
It does satisfy the dependency ranges: `jquery 3.2.1` satisfies `jquery >= 1.7` and `jquery 2.2.4` satisfies `jquery ^2.1.3`. `pickadate` should list `jquery` as a `peerDependency` if you want it to respect the project's jquery version requirement too.

**Comment by torgeir** [comment]
Thanks. Too fast, it being a plugin that makes sense that makes sense, I'll close this. Do you know what changed yarn's behavior in this regard?

**Comment by BYK** [comment]
@torgeir hey!

> Do you know what changed yarn's behavior from v0.27.0 to v0.27.1 in this regard?

This was right before I joined the project I guess but I'm assuming it was probably about a fix related how we resolve latest versions from ranges or may be related to our hoisting algorithm. We can dig up the exact commit which changed this behavior if you are interested tho :)

>Being a patch (or minor as you're <1) version I would not expect such changes, it broke hard on our end..

I totally agree, share the pain and apologize for the inconvenience we caused. We are working hard towards Yarn 1.0, where one of the major goals is to not make any breaking changes without major revision changes. We also want to put proper safeguards to protect against accidental changes and will also try to revert and release patch releases if/when we break this contract. Thanks a lot for bringing this up!
